src/main/java/org/logstash/filters/GeoIpFilter.java

```diff
@@ -0,0 +1,343 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.logstash.filters;
+
+import com.maxmind.db.CHMCache;
+import com.maxmind.db.InvalidDatabaseException;
+import com.maxmind.geoip2.exception.AddressNotFoundException;
+import com.maxmind.geoip2.exception.GeoIp2Exception;
+import com.maxmind.geoip2.model.CityResponse;
+import com.maxmind.geoip2.model.CountryResponse;
+import com.maxmind.geoip2.model.IspResponse;
+import com.maxmind.geoip2.record.*;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.logstash.Event;
+
+import com.maxmind.geoip2.DatabaseReader;
+import org.logstash.ext.JrubyEventExtLibrary.RubyEvent;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.*;
+
+public class GeoIpFilter {
+  private static Logger logger = LogManager.getLogger();
+  private static final String CITY_DB_TYPE = "GeoLite2-City";
+  private static final String COUNTRY_DB_TYPE = "GeoLite2-Country";
+  private static final String ISP_DB_TYPE = "GeoIP2-ISP";
+  
+  private final String sourceField;
+  private final String targetField;
+  private final Set<Fields> desiredFields;
+  private final DatabaseReader databaseReader;
+
+  public GeoIpFilter(String sourceField, String targetField, List<String> fields, String databasePath, int cacheSize) {
+    this.sourceField = sourceField;
+    this.targetField = targetField;
+    final File database = new File(databasePath);
+    try {
+      this.databaseReader = new DatabaseReader.Builder(database).withCache(new CHMCache(cacheSize)).build();
+    } catch (InvalidDatabaseException e) {
+      throw new IllegalArgumentException("The database provided is invalid or corrupted.", e);
+    } catch (IOException e) {
+      throw new IllegalArgumentException("The database provided was not found in the path", e);
+    }
+    this.desiredFields = createDesiredFields(fields);
+  }
+
+  private Set<Fields> createDesiredFields(List<String> fields) {
+    Set<Fields> desiredFields = EnumSet.noneOf(Fields.class);
+    if (fields == null) {
```

Comment:
also worth checking if fields.isEmpty() ? Or would we reject an empty fields list at configuration-time?